File: backend/app/models/task.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Enum, Float, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin
from app.models.enums import TaskPriority, TaskStatus, TaskType, VehicleType


class Task(Base, TimestampMixin):
    __tablename__ = "tasks"

    id: Mapped[int] = mapped_column(primary_key=True)
    title: Mapped[str] = mapped_column(String(160), index=True)
    description: Mapped[str] = mapped_column(Text)
    task_type: Mapped[TaskType] = mapped_column(Enum(TaskType, native_enum=False), index=True)
    region: Mapped[str] = mapped_column(String(80), index=True)
    district: Mapped[str] = mapped_column(String(80), index=True)
    location_name: Mapped[str] = mapped_column(String(120))
    lat: Mapped[float] = mapped_column(Float)
    lng: Mapped[float] = mapped_column(Float)
    priority: Mapped[TaskPriority] = mapped_column(Enum(TaskPriority, native_enum=False), index=True)
    scheduled_start: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    scheduled_end: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    status: Mapped[TaskStatus] = mapped_column(Enum(TaskStatus, native_enum=False), index=True)
    required_vehicle_type: Mapped[VehicleType | None] = mapped_column(Enum(VehicleType, native_enum=False), nullable=True)
    assigned_vehicle_id: Mapped[int | None] = mapped_column(ForeignKey("vehicles.id"), nullable=True)
    assigned_driver_id: Mapped[int | None] = mapped_column(ForeignKey("drivers.id"), nullable=True)

    assigned_vehicle = relationship("Vehicle", back_populates="tasks", foreign_keys=[assigned_vehicle_id])
    assigned_driver = relationship("Driver", back_populates="tasks", foreign_keys=[assigned_driver_id])
    alerts = relationship("Alert", back_populates="task")

File: flutter_app/lib/models/task.dart
enum TaskStatus {
  rejalashtirilgan,
  jarayonda,
  bajarildi,
  bekorQilindi,
}

enum TaskPriority {
  yuqori,
  orta,
  past,
}

class RecommendedVehicle {
  final String vehicleId;
  final String vehicleCode;
  final int score;
  final String reason;

  const RecommendedVehicle({
    required this.vehicleId,
    required this.vehicleCode,
    required this.score,
    required this.reason,
  });
}

class Task {
  final String id;
  final String title;
  final String type;
  final TaskPriority priority;
  final String region;
  final String district;
  final String scheduledStart;
  final String scheduledEnd;
  final TaskStatus status;
  final String? assignedVehicle;
  final String? assignedVehicleCode;
  final List<RecommendedVehicle> recommendedVehicles;
  final String description;
  final String createdAt;
  final double? lat;
  final double? lng;

  const Task({
    required this.id,
    required this.title,
    required this.type,
    required this.priority,
    required this.region,
    required this.district,
    required this.scheduledStart,
    required this.scheduledEnd,
    required this.status,
    this.assignedVehicle,
    this.assignedVehicleCode,
    required this.recommendedVehicles,
    required this.description,
    required this.createdAt,
    this.lat,
    this.lng,
  });

  String get statusLabel {
    switch (status) {
      case TaskStatus.rejalashtirilgan:
        return 'Rejalashtirilgan';
      case TaskStatus.jarayonda:
        return 'Jarayonda';
      case TaskStatus.bajarildi:
        return 'Bajarildi';
      case TaskStatus.bekorQilindi:
        return 'Bekor qilindi';
    }
  }

  String get priorityLabel {
    switch (priority) {
      case TaskPriority.yuqori:
        return 'Yuqori';
      case TaskPriority.orta:
        return "O'rta";
      case TaskPriority.past:
        return 'Past';
    }
  }
}
